doc/read-the-docs-site/adr/0005-inlining-saturated-functions.md

```diff
@@ -0,0 +1,206 @@
+# ADR 5: Inlining fully saturated functions
+
+Date: 2022-12
+
+## Authors
+
+Marty Stumpf <marty.stumpf@iohk.io>
+
+## Status
+
+Accepted
+
+## Context
+
+As part of our effort to increase script capacity,
+we take from [Plutonomy](https://github.com/well-typed/plutonomy),
+which gives 15-30% speedup on benchmarks.
+See [PLT-1044](https://input-output.atlassian.net/browse/PLT-1044).
+
+The goal is to integrate what we know already works there and
+also add optimizations we can do before UPLC (Plutonomy only applies optimization to UPLC).
+We currently have an inliner pass for PIR and UPLC.
+One of the optimization we can add is inlining of saturated functions in both PIR and UPLC.
+From our analysis, this is one of the optimizations that contributes to the speedup in Plutonomy.
+
+For example:
+
+```haskell
+let f x y = 1
+    y = f q
+in f a b 
+```
+
+we should inline `f` on line 3 (even though it’s used more than once), because it’s saturated,
+i.e. it has all its arguments, and we can reduce it.
+([PLT-1146](https://input-output.atlassian.net/browse/PLT-1146))
+
+To do this we need to do some arity analysis of functions,
+so we know how many term/type arguments they have.
+We also need to decide whether or not to inline based on the call site, which we don’t currently.
+
+This ADR records the details of the design decisions we made for the arity analysis of functions.
+It also lists the weaknesses of the approach and plans to mitigate them.
+The mechanics of whether or not to inline based on the call site will be covered in a separate ADR.
+
+## Decision
+
+For a first round of implementation, we are adding a simplified pass.
+We will refine the implementation over time.
+In each round, we benchmark and test and investigate the results.
+
+### The rationale: Counting lambdas
+
+**Arity** is relevant when we talk about a fully applied function.
+A fully applied function has an arity of 0.
+The arity of a function is the number of arguments it takes minus the number of arguments applied.
+E.g., `\x.\y.x+y` has an arity of 2. `(\x.\y.x+y) 1` has an arity of 1.
```

Comment:
I suppose the question is whether the goal is to inline fully reducible terms or not. Which "arity" should I be talking about?

Reply:
Just need a clarification: by definition `\x. \y. body` has aridy=1 (since we don't have multilambdas yet), but here we consider its arity=2 for the purpose of determining whether it is fully applied. Or maybe there's a different way of saying it without using "arity" at all.

Reply:
Agreed, I'll remove mentions of arity, like I said in [the other comment](https://github.com/input-output-hk/plutus/pull/5027#discussion_r1060763054).